README.rst
.. contents::

Introduction
====================

``collective.scriptedredirect`` allows you to write HTTP 302 Moved Temporary and HTTP 301 Moved Permanently
redirects for your `Plone CMS <http://plone.org>`_ site in Python code.

.. image:: https://travis-ci.org/collective/collective.scriptedredirect.png

Benefits
====================

* The redirect logic is front-end web server independent: no need to touch variouos configuration files of Apache, Varnish or Nginx)

* Python allows to write more complex logic for redirects easier - no regular expressions!

* Python scripts in Plone have access to more complete state information:
  user logged in status, permissions, etc.

Usage
====================

Plone 5.0 and 5.1 versions supported.

Installation
----------------

Add add-on in *buildout.cfg*::

    eggs =
        ...
        collective.scriptedredirect

Run buildout.

Install *Scripted redirects in Python* in Site Setup > Add-ons.

Doing redirects through the web
--------------------------------

Edit *redirect_handler* in Zope Management Interface in your site root.

.. image :: http://cloud.github.com/downloads/collective/collective.scriptedredirect/Screen%20Shot%202012-09-25%20at%201.28.18%20AM.png

In the case of accident use ``?no_redirect`` HTTP query parameter to override
the redirecter and fix your site.


Doing redirects through the web
--------------------------------

You can also register a browser view called ``redirect_handler``.
In this case, you write the redirect code in addon Python code
and not through the web.

``redirect_handler`` view is always preferred over ``redirect_handler`` script.

Example Python code in ``redirector.py``::

    class TestingRedirectHandler(object):
        """ Redirect handler registered as a ``redirect_handler`` Zope 3 <browser:page>
        """

        def __init__(self, context, request):
            self.context = context
            self.request = request

        def __call__(self, url, host, port, path):
            """ Handle redirects per site.

            :param path: Path as written in HTTP request (not site virtual path)

            :return: None if no redirect needed, otherwise a string full HTTP URL to the redirect target

            :raise: zExceptions.Redirect or other custom redirect exception if needed
            """

            # Simple example: always access site over www. domain prefix
            if not url.startswith("http://www."):
                return url.replace("http://", "http://www.")

            # Don't redirect if we are already using www. prefix
            return None

Example ZCML::

    <browser:page
        name="redirect_handler"
        for="Products.CMFCore.interfaces.ISiteRoot"
        layer="YOUR_ADDON_LAYER"
        class=".redirector.TestRedirectHandler"
        permission="zope.Public"
        />

Internals
====================

``collective.scriptedredirect`` hooks itself to Zope's pre-traversal hook and is
triggered before the request traverses into your Plone site in Zope application server.

Author
====================

Mikko Ohtamaa (`blog <https://opensourcehacker.com>`_, `Facebook <https://www.facebook.com/?q=#/pages/Open-Source-Hacker/181710458567630>`_, `Twitter <https://twitter.com/moo9000>`_, `Google+ <https://plus.google.com/u/0/103323677227728078543/>`_)





CHANGES.txt
Changelog
=========


2.0 (unreleased)
------------------

- Plone 5 compatibility
  [tomgross]
- Python 3 compatibility
  [karalics]


1.1.1 (2013-04-14)
------------------

- Fixed .rst title levels for PyPi [miohtama]


1.1 (2013-04-14)
----------------

- Added redirects via browser view support [miohtama]

- Testing, Travis CI support [miohtama]


1.0
----------------

- Initial development [miohtama]


src/collective/scriptedredirect/tests/test_view.py
# -*- coding: utf-8 -*-
import unittest

from zExceptions import Redirect

import zope.component
import zope.publisher.interfaces.browser
from zope.publisher.browser import BrowserView

from collective.scriptedredirect.testing import\
    COLLECTIVE_SCRIPTEDREDIRECT_FUNCTIONAL

from plone.testing.z2 import Browser


class TestingRedirectHandler(BrowserView):
    """ Redirect handler registered as a
       ``redirect_handler`` Zope 3 <browser:page>
    """

    def __call__(self, url, host, port, path):
        """
        :return: None if no redirect needed, otherwise a string
        full HTTP URL to the redirect target

        :raise: zExceptions.Redirect or other
        custom redirect exception if needed
        """

        # Simple example: always access site over www. domain prefix
        if not url.startswith("http://www."):
            return url.replace("http://", "http://www.")

        # Don't redirect if we are already using www. prefix
        return None


class TestRedirectView(unittest.TestCase):
    """
    Test redirect view.
    """
    layer = COLLECTIVE_SCRIPTEDREDIRECT_FUNCTIONAL

    def setUp(self):
        self.app = self.layer['app']
        self.portal = self.layer['portal']

        # Register view directly by pushing it to
        # zope.component multi-adapter registry

        zope.component.provideAdapter(
            # Our class
            factory=TestingRedirectHandler,
            # (context, request) layers for multiadapter lookup
            # We provide None as layers are not used
            adapts=(None, None),
            # All views are registered as IBrowserView interface
            provides=zope.publisher.interfaces.browser.IBrowserView,
            # View name
            name='redirect_handler')

        browser = Browser(self.portal)
        browser.handleErrors = False
        browser.raiseHttpErrors = False
        self.browser = browser

    def tearDown(self):
        """
        """
        # Dynamically unregister a view
        gsm = zope.component.getGlobalSiteManager()
        gsm.unregisterAdapter(
            factory=TestingRedirectHandler,
            required=(None, None),
            provided=zope.publisher.interfaces.browser.IBrowserView,
            name="redirect_handler"
        )

    def test_redirect_view(self):
        """ Test www redirect.

        View is registered through testing.zcml.
        """

        try:
            self.browser.open(self.portal.absolute_url())
            raise AssertionError("Should not be reached")
        except Redirect as e:
            # Plone test browser handlers
            # directs always as exceptions
            self.assertEqual(e.message, "http://www.nohost/plone")


src/collective/scriptedredirect/tests/test_redirect.py
# -*- coding: utf-8 -*-
import unittest

import transaction

from zExceptions import Redirect

from collective.scriptedredirect.testing import\
    COLLECTIVE_SCRIPTEDREDIRECT_FUNCTIONAL

from plone.testing.z2 import Browser


WWW_REDIRECT_SNIPPET = """
if not url.startswith("http://www."):
    return url.replace("http://", "http://www.")
"""

# Invalid Python code
BAD_SNIPPET = """
foobarbar
"""


class TestRedirectScript(unittest.TestCase):
    """
    Test redirect script.
    """
    layer = COLLECTIVE_SCRIPTEDREDIRECT_FUNCTIONAL

    def setUp(self):
        self.app = self.layer['app']
        self.portal = self.layer['portal']

        browser = Browser(self.app)
        browser.handleErrors = False
        browser.raiseHttpErrors = False
        self.browser = browser

    def set_redirect_script(self, payload):

        # Define the script parameters
        parameters = "url, host, port, path"

        script = self.portal.redirect_handler
        script.ZPythonScript_edit(parameters, payload)
        transaction.commit()

    def test_redirect_default_no_redirect(self):
        """ Test that the default installation doesn't produce redirects.
        """
        # Assert no error is risen
        self.browser.open(self.portal.absolute_url())
        self.assertEqual(self.browser.headers["status"], "200 Ok")

    def test_redirect_www(self):
        """ Set www redirect
        """
        self.set_redirect_script(WWW_REDIRECT_SNIPPET)

        try:
            self.browser.open(self.portal.absolute_url())
            raise AssertionError("Should not be reached")
        except Redirect as e:
            # Plone test browser handlers directs always as exceptions
            self.assertEqual(e.message, "http://www.nohost/plone")

    def test_bad_snippet(self):
        """ See that syntax error in the script does not cause problems.
        """
        self.set_redirect_script(BAD_SNIPPET)
        self.browser.open(self.portal.absolute_url())
        self.assertEqual(self.browser.headers["status"], "200 Ok")


src/collective/scriptedredirect/tests/test_installation.py
# -*- coding: utf-8 -*-
import unittest

from Products.CMFCore.utils import getToolByName

from collective.scriptedredirect.testing import\
    COLLECTIVE_SCRIPTEDREDIRECT_INTEGRATION


class TestInstallation(unittest.TestCase):

    layer = COLLECTIVE_SCRIPTEDREDIRECT_INTEGRATION

    def setUp(self):
        self.app = self.layer['app']
        self.portal = self.layer['portal']
        self.qi_tool = getToolByName(self.portal, 'portal_quickinstaller')

    def test_product_is_installed(self):
        """ Validate that our products GS profile has been run and the product
            installed
        """
        pid = 'collective.scriptedredirect'
        installed = [p['id'] for p in self.qi_tool.listInstalledProducts()]
        self.assertTrue(pid in installed,
                        'package appears not to have been installed')


src/collective/scriptedredirect/setuphandlers.py
# -*- coding: utf-8 -*-
from Products.PythonScripts.PythonScript import manage_addPythonScript

DEFAULT_REDIRECT_PY_CONTENT = """
if port not in (80, 443):
    # Don't kick in HTTP/HTTPS redirects if the site
    # is directly being accessed from a Zope front-end port
    return None

# Some default rules - comment out which ones you want to use.
# Use incoming 'url' and 'port' parameters to do the redirect.
# You return new redirect target as return value or
# directly raise an exception if special handling is needed

# Make sure that search engines and visitors access
# the site only WITH www. prefix
# if not url.startswith("http://www."):
#    return url.replace("http://", "http://www.")

# Make sure that search engines and visitors access
# the site only WITHOUT www. prefix
# if url.startswith("http://www."):
#    return url.replace("http://www.", "http://")

"""


def runCustomInstallerCode(site):
    """ Run custom add-on product installation code to modify Plone 
    site object and others Python scripts can be created by
    Products.PythonScripts.PythonScript.manage_addPythonScript

    http://svn.zope.org/Products.PythonScripts/trunk/src/Products/PythonScripts/PythonScript.py?rev=114513&view=auto  # noqa

    @param site: Plone site
    """

    # Create the script in the site root
    id = "redirect_handler"

    # Don't override the existing installation
    if id not in site.objectIds():
        manage_addPythonScript(site, id)
        script = site[id]

        # Define the script parameters
        parameters = "url, host, port, path"

        script.ZPythonScript_edit(parameters, DEFAULT_REDIRECT_PY_CONTENT)


def setupVarious(context):
    """
    @param context: Products.GenericSetup.context.DirectoryImportContext
    instance
    """

    # We check from our GenericSetup context whether we are running
    # add-on installation for your product or any other proudct
    if context.readDataFile('collective.scriptedredirect.marker.txt') is None:
        # Not our add-on
        return

    portal = context.getSite()

    runCustomInstallerCode(portal)


src/collective/scriptedredirect/redirect.py
# -*- coding: utf-8 -*-
"""

    Call a custom TTW script and allow it to handle redirects.


    Use Zope Management Interface to add a ``Script (Python)`` item
    named ``redirect_handler`` to your site root - you can edit this
    script in fly to change the redirects.

    * Redirect script must contain ``url`` in its parameter list

"""


import logging
try:
    from urlparse import urlparse
except ImportError:
    import urllib.parse as urlparse
# Really old old stuff
from zExceptions import Redirect
from ZODB.POSException import ConflictError

from zope.component import queryMultiAdapter

logger = logging.getLogger("redirect")


def get_redirect_handler_for_site(site, request):
    """ Get view or script handling the redirect.

    :return: callable or None
    """

    view = queryMultiAdapter((site, request), name="redirect_handler")
    if view:
        return view

    # Check if we have a redirect handler script in the site root
    if "redirect_handler" in site:
        return site["redirect_handler"]

    return None


def check_redirect(site, event):
    """
    Check if we have a custom redirect script in Zope application server root.

    If we do then call it and see if we get a redirect.

    The script itself is TTW Python script which may return
    string in the case of redirect or None if no redirect is needed.

    For more examples, check

    http://svn.zope.org/Zope/trunk/src/Zope2/App/tests/testExceptionHook.py?rev=115555&view=markup   # noqa
    """
    request = event.request

    url = request["ACTUAL_URL"]

    parts = urlparse(url)
    netloc = parts.netloc.split(":")
    host = netloc[0]

    path = parts.path

    if len(netloc) > 1:
        port = int(netloc[1])
    else:
        if parts.scheme == "https":
            port = 443
        else:
            port = 80

    if "no_redirect" in request.form:
        # Use no_redirect query parameter to disable this behavior in the case
        # you mess up with the redirect script
        return

    # Check if we have a redirect handler script in the site root
    handler = get_redirect_handler_for_site(site, request)
    if handler:

        try:
            # Call the script and get its output
            value = handler(url=url, host=host, port=port, path=path)

            if value is not None and value.startswith("http"):
                # Trigger redirect, but only if the output value looks sane
                raise Redirect(value)
        except ConflictError:
            # Zope 2 retry exception
            raise
        except Redirect:
            # Redirect exceptions are the only legal ones
            # from above logic
            raise
        except Exception as e:
            # No silent exceptions plz
            logger.error("Redirect exception for URL:" + url)
            logger.exception(e)
            return
